pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.type;
+
+import java.util.Map;
+import org.apache.calcite.jdbc.JavaTypeFactoryImpl;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * Extends Java-base TypeFactory from Calcite.
+ */
+public class TypeFactory extends JavaTypeFactoryImpl {
+  private final RelDataTypeSystem _typeSystem;
+
+  public TypeFactory(RelDataTypeSystem typeSystem) {
+    _typeSystem = typeSystem;
+  }
+
+  public RelDataType createRelDataTypeFromSchema(Schema schema) {
+    Builder builder = new Builder(this);
+    for (Map.Entry<String, FieldSpec> e : schema.getFieldSpecMap().entrySet()) {
+      builder.add(e.getKey(), toRelDataType(e.getValue()));
+    }
+    return builder.build();
+  }
+
+  private RelDataType toRelDataType(FieldSpec fieldSpec) {
```

Comment:
How do we handle/factor array / MV ?

Reply:
calcite supports Map, Array and Struct type. but we are throwing here until operator support is added.